### PR TITLE
Add support for blueprint-level cli commands

### LINKED ISSUE #1357: Support for @blueprint.cli.command
It would be great to have access to the functionality of `app.cli.command` via blueprints - i.e. something like:

``` python
blueprint = Blueprint('myblueprint', __name__,)

@blueprint.cli.command()
def doblueprintstuff():
    """Initialize the database."""
    #....
```

It would help a lot when building modular applications.

### PR DESCRIPTION
**Pull request by ThiefMaster**
Still WIP, but that way it can already be reviewed and commented on.

Tests are still missing and so is a `Blueprint.app_cli` which registers app-level commands from a blueprint (which I think makes sense depending on how one uses blueprints).

When finished, this will close #1357

### CODE REVIEW

File: flask/blueprints.py [hunk 1]
```diff
@@ -105,6 +106,19 @@ def __init__(self, name, import_name, static_folder=None,
         if url_defaults is None:
             url_defaults = {}
         self.url_values_defaults = url_defaults
+        if cli_group_name is None:
+            cli_group_name = name
+        self.cli_group_name = cli_group_name
+        if cli_group_desc is None:
+            cli_group_desc = 'Commands provided by {0}.'.format(self.name)
```
Comment (untitaker) [inline flask/blueprints.py]: I think given the context, it should be clarified that this is a blueprint.

Comment (ThiefMaster) [inline flask/blueprints.py]: I initially had a longer description (*"Commands provided by the ... blueprint"*), but click ended up truncating it...

Comment (untitaker) [inline flask/blueprints.py]: "Commands by the {} blueprint" maybe?

Comment (ThiefMaster) [inline flask/blueprints.py]: good idea; that way only the "blueprint" is truncated if the name is too long, and at least with my test blueprint ("yellowprint") nothing was truncated at all

**Comment by untitaker** [comment]
Maybe registering global commands should be done like with routes: During registration, provide a prefix, otherwise it's global by default.

**Comment by ThiefMaster** [comment]
Sounds fine, but that might get a bit ugly when you end up using the same group/prefix in two blueprints. Each of them *must* create the group immediately (since you can nest them and recording all the nested calls would be somewhat of an overkill), so they'd have to be merged during registration.

After looking into `click.Group` that might not be such a big problem though, since I'd just have to merge the `commands` dict from the new group with the existing group's command dict...

**Comment by untitaker** [comment]
Sounds tricky, but having `app_cli` before registering to the app sounds equally tricky.

**Comment by ThiefMaster** [comment]
Managed to get it working. What I still don't like that much is the fact that you can only specify additional arguments such as the help text for groups created using the default cli_group of the the blueprint. Adding another kwarg to the decorators is not really a good idea since the group description is not really related to a command.

Maybe the better solution would be getting rid of the `group` kwarg and simply creating the group using the decorator and creating its commands using the decorator provided by click.Group.command...

**Comment by ThiefMaster** [comment]
Simplified it a lot in the second commit. I think it's more straightforward and especially closer to the regular `click` API like that.

**Comment by ThiefMaster** [comment]
I guess I still need to merge groups if two blueprints define the same one. Going to have a look at that after work...

**Comment by untitaker** [comment]
Hey @ThiefMaster what's the status of this?

**Comment by ThiefMaster** [comment]
It's not really clear what's the best way to implement this so it probably makes more sense to close this for now...

**Pull request closed without merging**